User implementation

Guptacos opened the pull request.

Guptacos committed:
Start User class implementation

- Implement init, str, and player methods
- Set up testing file with tests for user.player

Guptacos committed:
Some preliminary User implementation

- Add dummy classes for artist, album, etc. so code runs
- Write and test init, str, _update_internal, top
- Remove 'Response'
- Update documentation that limit must be non-negative

Guptacos committed:
Add and test User.recently_played

Guptacos committed:
Add and test User.get_playlists

Guptacos committed:
Create pagination helper function

User._paginate_get paginates requests to spotify to generate more
results than can be gotten from just one request.

Guptacos committed:
Add and test User.create_playlist

Guptacos committed:
Add and test User.get_following

Guptacos committed:
Add and test User.is_following

Guptacos committed:
Add and test has_saved

Also reformat code, taking repeat functionality and boxing into helper
functions.

Guptacos committed:
Add and test User.get_saved

Guptacos committed:
Document auth requirements

Add documentation for follow, unfollow, save, and remove.

Guptacos committed:
Add tests for the User class

This file was renamed and deleted in commit history during a recent
rebase. This commit reintroduces the file to version control.

Guptacos committed:
Intermediate commit

This is a staging ground for changes that I've been working on. The
tests are partially mocked, user is partially implemented, and the
linter partially passes.

Guptacos committed:
Remove typehints and satisfy linter

- Type hinting was creating odd import errors. Since types are specified
  in the funciton docstring, remove formal type hinting and type
  checking.
- Update user.py and test_user.py to satisfy pylint

Guptacos committed:
Move import statements to compile

To solve the circular dependency issue, remove static type hinting and
move imports of other Spotify objects (i.e. Track, Album, Artist,
Playlist, Player, User) to the bottom of the file.

Guptacos committed:
Add more dummy data to artists, albums, and tracks

Guptacos committed:
Update user.py as per new style guidelines

Also pluralize contants.

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,770 @@
+                 user_info=None):
+        """
+        Args:
+            session: a Spotifython instance
```
I'm not sure if this is clear that it is an instance of the session class.

Guptacos replied:
Changed comment to Session

Guptacos committed:
Temporary intermediate commit

Guptacos committed:
Finish implementation of User class

Still need to iron out some of the test cases.

Guptacos committed:
Add dunder methods to User class

Add eq, ne, and hash.

Guptacos committed:
Remove self._info and _update_internal

The only static information that is cached is the id. As such it makes
sense to remove the rest of the unnecessary data.

Guptacos committed:
Clean up documentation / input validation for User

Guptacos committed:
Round out User class and tests

User.is_following and User.get_following still need to be tested.

spenceryu commented on tests/test_user.py:
```diff
@@ -0,0 +1,507 @@
+        self.assertIsInstance(user.player(), Player)
+
+
+    # User.user_id
```
.spotify_id

Guptacos replied:
fixed

spenceryu commented on spotifython/constants.py:
```diff
@@ -25,3 +25,5 @@
 DEFAULT_REQUEST_TIMEOUT = 10 # in seconds
 SPOTIFY_PAGE_SIZE = 50
+SPOTIFY_MAX_PLAYLISTS = 100000
+SPOTIFY_MAX_LIB_SIZE = 10000
```
SPOTIFY_MAX_LIBRARY_SIZE?

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+
+        Args:
+            top_type: get top items of this type. One of:
+                sp.ARTISTS
```
rename to const.ARTISTS

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+        Args:
+            top_type: get top items of this type. One of:
+                sp.ARTISTS
+                sp.TRACKS
```
const.TRACKS

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+                sp.TRACKS
+            limit: (int) max number of items to return. Must be positive.
+            time_range: get top items for this time range. One of:
                 sp.LONG (several years)
```
const.LONG/MEDIUM/SHORT

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
-            Failure: None
+        Returns:
+            A list of artists or a list of tracks, depending on top_type. Could
+            be empty.
```
Could return an empty list.

Guptacos replied:
fixed

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+                        self._session,
+                        limit=limit,
+                        return_class=return_class,
+                        endpoint=Endpoints.USER_TOP % endpoint_type,
```
Endpoints: I think we should standardize to using "{}".format(arg) rather than % because it is a newer format.

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+            name: (str) The name for the new playlist. Does not need to be
+                unique; a user may have several playlists with the same name.
+            visibility: how other users interact with this playlist. One of:
                     sp.PUBLIC: publicly viewable, not collaborative
```
const.

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+        if limit <= 0 or limit > const.SPOTIFY_MAX_PLAYLISTS:
+            raise ValueError(limit)
+
+        endpoint = Endpoints.USER_GET_PLAYLISTS % self.spotify_id()
```
"{}".format()

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+        response_json, status_code = utils.request(
+            self._session,
+            request_type=const.REQUEST_POST,
+            endpoint=Endpoints.USER_CREATE_PLAYLIST % self.spotify_id(),
```
"{}".format()

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+
+        Args:
+            follow_type: one of:
+                sp.ARTISTS
```
const.

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+
+        Args:
+            saved_type: one of:
+                sp.ALBUMS
```
const

spenceryu commented on spotifython/user.py:
```diff
@@ -1,371 +1,771 @@
+                Used for track relinking:
+                https://developer.spotify.com/documentation/general/guides/track-relinking-guide/
+
+                if sp.TOKEN_REGION (default) is given, will use appropriate
```
const

spenceryu reviewed (commented):
lgtm but there's just some comments that need to be updated to reflect our naming conventions for constants and some minor comment nits

spenceryu reviewed (dismissed):
lgtm but there's just some comments that need to be updated to reflect our naming conventions for constants and some minor stylistic nits

Guptacos committed:
Satisfy linter and clear TODOS for User class

Guptacos committed:
Fix Spencers suggestions / satisfy linter for User

Guptacos committed:
Fix import-error in spotifython/utils.py

Guptacos committed:
Update documentation for User

Change formatting to fit with Sphinx requirements.

rmannava commented on spotifython/user.py:
```diff
@@ -1,371 +1,790 @@
-        # note to self: init self.player
-        pass
+    def __init__(self, session, info):
+        """
```
should include disclaimer about usage

rmannava commented on spotifython/constants.py:
```diff
@@ -1,3 +1,6 @@
+""" Constants and enums used by the spotifython library.
+"""
```
have this set of quotes at the end of the first line?

rmannava commented on spotifython/user.py:
```diff
@@ -1,371 +1,790 @@
+        """
+        # Validate follow_type
+        if follow_type not in [const.ARTISTS, const.PLAYLISTS]:
+            raise TypeError(follow_type)
```
should state the accepted types in error

Guptacos replied:
I agree, but this should be consistent throughout all our code. I'll add a top level TODO in the backlog.

rmannava commented on spotifython/user.py:
```diff
@@ -1,371 +1,790 @@
+        # Validate limit
+        if limit is not None:
+            if limit <= 0:
+                raise ValueError(limit)
+            if follow_type == const.PLAYLISTS and limit > const.MAX_PLAYLISTS:
+                raise ValueError(limit)
```
some descriptive error message

rmannava commented on tests/README.md:
```diff
@@ -14,3 +14,16 @@ Take a look at template.py, and use that as a base to create unit tests for your
+
+
+## Dummy data
+Data in the dummy\_data directory can be used for mocking the spotify api.
+
+- Tracks and albums were pulled from the public Spotify API.
+- Users were generated using a script (since Spotify provides no easy way to get
+  random user ids) and a default template.
+- Artists were taken from my top artists and some recommended from browse
+- Playlists taken from https://api.spotify.com/v1/browse/featured-playlists, and
+  augmented using the public playlists in my library.
+
+Data can be easily accessed using help\_lib.py
```
directories forward slash?

Guptacos replied:
Backslash is for formatting; otherwise markdown reads the _ as used for italics / bold

rmannava commented on tests/dummy_data/generate.py:
```diff
@@ -1,48 +1,72 @@
+'''
+Since you can't get random users from the Spotify interface, randomly generate
+some fake users for testing.
+
+Template is taken from Spotify User ODM
+
+Usage:
+`python generate.py _num_entries_` to print to the console
+`python generate.py _num_entries_ > file.json` to save to a file
+
+'''
```
double quote """

rmannava commented on tests/test_user.py:
```diff
@@ -0,0 +1,507 @@
+''' Tests for the User class
+
+Note: these tests are not exhaustive and could always be improved. Since the
+Spotify REST api is mocked, if it's functionality ever changes these tests may
+become obsolete.
+
+Last updated: May 14, 2020
+
+'''
```
double quote """

rmannava commented on tests/test_user.py:
```diff
@@ -0,0 +1,507 @@
+''' Tests for the User class
+
+Note: these tests are not exhaustive and could always be improved. Since the
+Spotify REST api is mocked, if it's functionality ever changes these tests may
```
it's -> its

Guptacos committed:
Fix review nits in User
